packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -174,6 +179,126 @@ class Plugin {
     );
   }
 
+  /// Queries the `pubpsec` and returns the `platforms` sub map.
+  ///
+  /// Throws ToolExit if cannot find the `platforms:` map.
+  static YamlMap getPlatformsYamlMap(YamlMap pubspec) {
+    if (pubspec == null) {
+       throwToolExit('Invalid pubspec.yaml', exitCode: 2);
+    }
+    final YamlMap flutterConfig = pubspec['flutter'] as YamlMap;
+    if (flutterConfig == null) {
+      throwToolExit('Did not find valid `flutter` map in pubspec.yaml', exitCode: 2);
+    }
+    final YamlMap pluginConfig = flutterConfig['plugin'] as YamlMap;
+    if (pluginConfig == null) {
+      throwToolExit('Did not find valid `plugin` map in pubspec.yaml', exitCode: 2);
+    }
+    if (pluginConfig['platforms'] == null || pluginConfig['platforms'] is! YamlMap) {
+      throwToolExit(_invalidPlatformsErrorMessage, exitCode: 2);
+    }
+
+    return pluginConfig['platforms'] as YamlMap;
+  }
+
+  /// Updates the pubspec.yaml under `projectDir`, adding each platform in `platforms` under `plugin: platforms:`
+  ///
+  /// If the dummy platform `some_platform` exists, remove it.
+  /// For example, if the original `plugin` map is:
+  /// ```
+  ///   flutter:
+  ///      plugin:
+  ///        platforms:
+  ///          some_platform:
+  ///            pluginClass: somePluginClass
+  /// ```
+  /// and the `platforms` is specified as `['ios','android'], the yaml map becomes:
+  /// ```
+  ///    flutter:
+  ///      plugin:
+  ///        platforms:
+  ///          android:
+  ///            package: <androidIdentifier>
+  ///            pluginClass: <pluginClass>
+  ///          ios:
+  ///            pluginClass: <pluginClass>
+  /// ```
+  ///
+  /// This method doesn't support the legacy plugin pubspec format and will throw error if try to update a pubspec.yaml that uses the legacy format.
+  static Future<void> updatePubspecWithPlatforms(String projectDir, List<String> platforms, String pluginClass, String androidIdentifier) async {
+    if (platforms.isEmpty) {
+      return;
+    }
+
+    final String pubspecPath = globals.fs.path.join(projectDir, 'pubspec.yaml');
+    if (!globals.fs.file(pubspecPath).existsSync()) {
+      return;
+    }
+    final YamlMap pubspec = loadYaml(globals.fs.file(pubspecPath).readAsStringSync()) as YamlMap;
```

Comment:
Can we move the parsing and validation in this method by using FlutterManifest?
```dart
final FlutterManifest manifest = FlutterManifest.createFromPath(pubspecPath, fileSystem: globals.fs, logger: globals.logger);
```
Then move `getPlatformsYamlMap` logic into `FlutterManifest` as something like this:
```dart
  List<Map<String, dynamic>> get supportedPlatforms {
    if (isPlugin) {
      final YamlMap plugin = _flutterDescriptor['plugin'] as YamlMap;
      if (plugin.containsKey('platforms')) {
        return platforms = plugin['platforms'] as YamlMap;
      }
    }
    return <String>[];
  }
```
And put any missing type logic (if any) into `_validateMultiPlatformYaml()`.

(bonus, then androidPackage can become:)
```dart
  String get androidPackage {
    if (isModule) {
      return _flutterDescriptor['module']['androidPackage'] as String;
    }
    if (supportedPlatforms.containsKey('android')) {
      return supportedPlatforms['android']['package'] as String;
    }
    return null;
  }
```